control/gps.py

```diff
@@ -37,9 +66,17 @@ def __repr__(self):
     def read(self):
         """Returns a GpsReading object with the values supplied"""
         msg = None
-        # TODO: Add try except chain to catch any dirty GPS output exceptions
-        # TODO: Add timeout for reads using signal library
-        while not isinstance(msg, pynmea2.GGA):
-            msg = pynmea2.parse(self.ser.readline())
+        tries = 4
+        for attempt in range(1, tries + 1):
+            try:
+                while not isinstance(msg, pynmea2.GGA):
```

Comment:
if we don't have a lock does the data nmea sentence still define the type? Like would there be a situation where the nmea sentence is never a GGA...not saying we aren't getting data from the gps, but getting data and its just all null data. Tryna imagine a situation where this while loop never breaks

Reply:
I believe that the gps will alway output a GGA message even if there's no lock. The unit test 'test_gps_satellite_no_lock_msg' actually tests this. It'll parse it and return a 0, 0 for lat and lon. So the loop will break either by a corrupt message or a some form of a GGA message. If no serial data is coming through then the timeout should handle that as well.